core/src/main/java/io/ddf/DDFCache.java

```diff
@@ -2,52 +2,90 @@
+
+  private LoadingCache<UUID, DDF> mDDFCache;
+
+  private DDFManager mDDFManager;
+
+  public DDFCache(DDFManager manager) {
+    mDDFManager = manager;
+    /**
+     * Maximum number of DDFs allowed in cache, DDFs will be evicted by an approximate LRU algorithm
+     */
+    Long maxNumberOfDDFs = Long.valueOf(Config.getGlobalValue(Config.ConfigConstant.MAX_NUMBER_OF_DDFS_IN_CACHE));
+
+    /**
+     * The number of seconds that the DDF will live in cache since it last used(meaning inserted or accessed)
+     */
+    Long timeToIdle = Long.valueOf(Config.getGlobalValue(Config.ConfigConstant.DDF_TIME_TO_IDLE_SECONDS));
+
+    mLog.info(String.format("Maximum number of ddfs in cache %s", maxNumberOfDDFs));
+    mLog.info(String.format("DDF's time to idle = %s seconds", timeToIdle));
+    mDDFCache = CacheBuilder.newBuilder().
+        maximumSize(maxNumberOfDDFs).expireAfterAccess(timeToIdle, TimeUnit.SECONDS).
+        recordStats().removalListener(new DDFRemovalListener())
+        .build(new CacheLoader<UUID, DDF>() {
+      @Override public DDF load(UUID uuid) throws Exception {
+        try {
+          mLog.info(String.format("restoring ddf %s", uuid));
+          return mDDFManager.restoreDDF(uuid);
+        } catch (Exception e) {
+          mLog.error(String.format("Error restoring DDF %s, error = %s", uuid, ExceptionUtils.getStackTrace(e)));
+          throw new DDFException(String.format("DDF with uuid %s does not exist", uuid));
+        }
+      }
+    });
+  }
+
+  private boolean containsDDF(UUID uuid) {
+    return mDDFCache.asMap().containsKey(uuid);
+  }
 
-  private Map<UUID, DDF> mDDFs = new ConcurrentHashMap<UUID, DDF>();
   private Map<String, UUID> mUris = new ConcurrentHashMap<String, UUID>();
 
   public void addDDF(DDF ddf) throws DDFException {
-    mDDFs.put(ddf.getUUID(), ddf);
+    if(ddf.getUUID() == null) {
+      throw new DDFException("uuid is null");
+    }
+    mDDFCache.put(ddf.getUUID(), ddf);
+  }
+
+  public CacheStats getCacheStats() {
+    return mDDFCache.stats();
   }
 
   public void removeDDF(DDF ddf) throws DDFException {
-    mDDFs.remove(ddf.getUUID());
+    mDDFCache.invalidate(ddf.getUUID());
     if(ddf.getUri() != null) {
       mUris.remove(ddf.getUri());
     }
   }
 
   public DDF[] listDDFs() {
-    return this.mDDFs.values().toArray(new DDF[] {});
+    return mDDFCache.asMap().values().toArray(new DDF[]{});
   }
 
-  public DDF getDDF(UUID uuid) throws DDFException {
-    DDF ddf = mDDFs.get(uuid);
-    if(ddf == null) {
-      throw new DDFException(String.format("Cannot find ddf with uuid %s", uuid));
-
-    } else {
-      return ddf;
-    }
-  }
-
-  public boolean hasDDF(UUID uuid) {
-    DDF ddf = mDDFs.get(uuid);
-    return ddf != null;
+  public DDF getDDF(UUID uuid) {
+    return mDDFCache.getUnchecked(uuid);
   }
 
   public DDF getDDFByName(String name) throws DDFException {
-    for(DDF ddf: mDDFs.values()) {
+    for(DDF ddf: this.listDDFs()) {
```

Comment:
4000 items is not that big, I think we can go on with this until we see performance degrades.
Using another hashmap will add the burden of maintaining consistency between the hashmap and mDDFCache, which may be tricky because this class will be used in different threads.